samples/drivers/i2c/rtio_loopback/src/main.c

```diff
@@ -0,0 +1,402 @@
+	 * so we add them to the read SQE.
+	 */
+	rd_sqe->iodev_flags = RTIO_IODEV_I2C_STOP | RTIO_IODEV_I2C_RESTART;
+
+	/*
+	 * With the two SQEs of the sample_rtio context prepared, we call
+	 * rtio_submit() to have them executed. This call will execute all
+	 * prepared SQEs.
+	 *
+	 * In this case, we wait for the two SQEs to be completed before
+	 * continuing, similar to calling i2c_transfer().
+	 */
+	ret = rtio_submit(&sample_rtio, 2);
+	if (ret) {
+		return -EIO;
+	}
+
+	/*
+	 * Since the RTIO SQEs are executed in the background, we need to
+	 * get the CQE after and check its results. Since we configured the
+	 * write and read SQEs as a single transaction, only one CQE is
+	 * generated which includes both of them. If we had chained them
+	 * instead, one CQE would be created for each of them.
+	 */
+	wr_rd_cqe = rtio_cqe_consume(&sample_rtio);
+	if (wr_rd_cqe->result) {
+		return -EIO;
+	}
+
+	/* Release the CQE after having checked its result. */
+	rtio_cqe_release(&sample_rtio, wr_rd_cqe);
+	return 0;
+}
+
+static void rtio_write_read_done_callback(struct rtio *r, const struct rtio_sqe *sqe, void *arg0)
+{
+	struct k_sem *sem = arg0;
+
+	/* Signal write and read SQEs completed */
+	k_sem_give(sem);
+}
+
+static int sample_rtio_write_read_async(void)
+{
+	struct rtio_sqe *wr_sqe, *rd_sqe, *cb_sqe;
+	struct rtio_cqe *wr_rd_cqe;
+	int ret;
+
+	/* See sample_rtio_write_read() */
+	wr_sqe = rtio_sqe_acquire(&sample_rtio);
+	rtio_sqe_prep_write(wr_sqe,
+			    &sample_rtio_iodev,
+			    0,
+			    sample_write_data,
+			    sizeof(sample_write_data), NULL);
+	wr_sqe->flags |= RTIO_SQE_TRANSACTION;
+
+	/* See sample_rtio_write_read() */
+	rd_sqe = rtio_sqe_acquire(&sample_rtio);
+	rtio_sqe_prep_read(rd_sqe,
+			    &sample_rtio_iodev,
+			    0,
+			    sample_read_buf,
+			    sizeof(sample_read_buf), NULL);
+	rd_sqe->iodev_flags = RTIO_IODEV_I2C_STOP | RTIO_IODEV_I2C_RESTART;
+
+	/*
+	 * The next SQE is a callback which we will use to signal that the
+	 * write and read SQEs have completed. It has to be executed only
+	 * after the write and read SQEs, which we configure by setting the
+	 * RTIO_SQE_CHAINED flag.
+	 */
+	rd_sqe->flags |= RTIO_SQE_CHAINED;
+
+	/*
+	 * Prepare the callback SQE. The SQE allows us to pass an optional
+	 * argument to the handler, which we will use to store a pointer to
+	 * the binary semaphore we will be waiting on.
+	 */
+	cb_sqe = rtio_sqe_acquire(&sample_rtio);
+	rtio_sqe_prep_callback_no_cqe(cb_sqe,
+				      rtio_write_read_done_callback,
+				      &sample_write_read_sem,
+				      NULL);
+
+	/*
+	 * Submit the SQEs for execution, without waiting for any of them
+	 * to be completed. We use the callback to signal completion of all
+	 * of them.
+	 */
+	ret = rtio_submit(&sample_rtio, 0);
+	if (ret) {
+		return -EIO;
+	}
+
+	/* We wait for the callback which signals RTIO transfer has completed */
+	ret = k_sem_take(&sample_write_read_sem, SAMPLE_TIMEOUT);
```

Comment:
rtio_cqe_consome_block() does much the same thing, could easily add a variant of the function with a timeout if you like and would be accepted